File: bms/go/go-gt/heap_down/heap_down.go
package test

var ()

const ()

func (h Heap) down(i int64) {
	for {
		left := 2*i + 1
		if left >= h.N {
			break
		}
		j := left
		if right := left + 1; right < h.N && !h.less(left, right) {
			j = right
		}
		if h.less(i, j) {
			break
		}
		h.swap(i, j)
		i = j
	}
}

func (h Heap) less(a, b int64) bool {
	i, j := h.I[a], h.I[b]
	return h.W[i] < h.W[j]
}

func (h Heap) swap(a, b int64) {
	i, j := h.I[a], h.I[b]
	h.I[a], h.I[b] = h.I[b], h.I[a]
	h.A[i], h.A[j] = b, a
}

type Heap struct {
	N int64
	I []int64
	A []int64
	W []int64
}

File: translations_only/fail/go/go-gt/heap_down/gemini/Hinted/heap_down.rs

fn down(h: &mut Heap, mut i: i64) {
    loop {
        let left = 2 * i + 1;
        if left >= h.N {
            break;
        }
        let mut j = left;
        if let Some(right) = left.checked_add(1) {
            if right < h.N && !less(h, left, right) {
                j = right;
            }
        }
        if less(h, i, j) {
            break;
        }
        swap(h, i, j);
        i = j;
    }
}

fn less(h: &Heap, a: i64, b: i64) -> bool {
    let i = h.I[a as usize];
    let j = h.I[b as usize];
    h.W[i as usize] < h.W[j as usize]
}

fn swap(h: &mut Heap, a: i64, b: i64) {
    let i = h.I[a as usize];
    let j = h.I[b as usize];
    h.I[a as usize] = h.I[b as usize];
    h.I[b as usize] = i;
    h.A[i as usize] = b;
    h.A[j as usize] = a;
}

struct Heap {
    N: i64,
    I: Vec<i64>,
    A: Vec<i64>,
    W: Vec<i64>,
}
